packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -650,9 +650,19 @@ mixin LocalHistoryRoute<T> on Route<T> {
 }
 
 class _DismissModalAction extends DismissAction {
+  _DismissModalAction(this.context);
+
+  final BuildContext context;
+
+  @override
+  bool isEnabled(DismissIntent intent) {
+    final ModalRoute<dynamic> route = ModalRoute.of<dynamic>(context)!;
+    return route.barrierDismissible;
+  }
+
   @override
   Object invoke(DismissIntent intent) {
-    return Navigator.of(primaryFocus!.context!)!.maybePop();
+    return Navigator.of(context)!.maybePop();
```

Comment:
The problem with this is that if the action is created above the navigator, or in the scope of the wrong navigator, it will do some very unexpected things. But I suppose, since this is a private action, you have plenty of control over that.

If this were a public action, you'd want to listen to focus changes by listening to the FocusManager,but you can't do that here, because the `Action` isn't a widget that can register/unregister with the FocusManager.

I actually think that the `Actions` widget itself needs to listen to focus changes and rebuild, since it is quite common for the state of an action to change as a result of focus changes. Focus changes aren't that often, so it wouldn't really be a performance issue.

(To be clear: I don't think you need to change anything here)